engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -0,0 +1,643 @@
+import io.deephaven.engine.table.impl.select.MatchPairFactory;
+import io.deephaven.engine.table.impl.sources.RedirectedColumnSource;
+import io.deephaven.engine.table.impl.sources.ReinterpretUtils;
+import io.deephaven.engine.table.impl.util.SingleValueRowRedirection;
+import io.deephaven.engine.table.impl.util.WritableRowRedirection;
+import io.deephaven.engine.table.impl.util.WritableSingleValueRowRedirection;
+import io.deephaven.engine.updategraph.NotificationQueue;
+import io.deephaven.internal.log.LoggerFactory;
+import io.deephaven.io.logger.Logger;
+import io.deephaven.util.annotations.TestUseOnly;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static io.deephaven.engine.rowset.RowSequence.NULL_ROW_KEY;
+import static io.deephaven.engine.table.impl.MultiJoinModifiedSlotTracker.*;
+
+public class MultiJoinTableImpl implements MultiJoinTable {
+
+    static final int KEY_COLUMN_SENTINEL = -2;
+    private final Table table;
+
+    @TestUseOnly
+    static MultiJoinTableImpl of(@NotNull final JoinControl joinControl,
+            @NotNull final MultiJoinInput... multiJoinInputs) {
+        return QueryPerformanceRecorder.withNugget("multiJoin",
+                () -> new MultiJoinTableImpl(joinControl, multiJoinInputs));
+    }
+
+    static MultiJoinTableImpl of(@NotNull final MultiJoinInput... joinDescriptors) {
+        return QueryPerformanceRecorder.withNugget("multiJoin",
+                () -> new MultiJoinTableImpl(new JoinControl(), joinDescriptors));
+    }
+
+    /**
+     * Get the output {@link Table table} from this multi-join table.
+     *
+     * @return The output {@link Table table}
+     */
+    public Table table() {
+        return table;
+    }
+
+    @Override
+    public Collection<String> keyColumns() {
+        return null;
+    }
+
+    private MultiJoinTableImpl(@NotNull final JoinControl joinControl,
+            @NotNull final MultiJoinInput... multiJoinInputs) {
+        table = multiJoin(joinControl, multiJoinInputs);
+    }
+
+    private Table multiJoin(@NotNull final JoinControl joinControl,
+            @NotNull final MultiJoinInput... multiJoinInputs) {
+        final TObjectIntHashMap<String> usedColumns =
+                new TObjectIntHashMap<>(multiJoinInputs[0].columnsToAdd().length, 0.5f, -1);
+        final List<String> expectedLeftMatches =
+                ColumnName.names(JoinMatch.lefts(Arrays.asList(multiJoinInputs[0].columnsToMatch())));
+        if (expectedLeftMatches.size() == 0) {
+            return doMultiJoinZeroKey(multiJoinInputs);
+        }
+
+        final Set<String> keyColumnNames = new LinkedHashSet<>(expectedLeftMatches);
+        for (final String keyColumn : expectedLeftMatches) {
+            usedColumns.put(keyColumn, KEY_COLUMN_SENTINEL);
+        }
+        final ChunkType[] expectedChunkTypes = Arrays.stream(getKeySources(multiJoinInputs[0]))
```

Comment:
You might want to do:
1. A single pass over the inputs to get *all* your match columns.
2. A single pass over the inputs to get *all* your addition columns.